File: lib/analysis/dfa.rb
# typed: strict
require "sorbet-runtime"
require_relative "analysis"
require_relative "cfg"
require_relative "cfg_facts"

# A DFA is a generic framework for a data-flow analysis.
class Analysis::DFA
  extend T::Sig
  extend T::Generic

  include Analysis

  Domain = type_member {{ upper: Object }}

  # The Direction enum specifies if a DFA runs forwards or backwards.
  class Direction < T::Enum
    enums do
      Forwards = new
      Backwards = new
    end
  end

  sig { params(direction: Direction,
               boundary: T::Set[Domain],
               init: T::Set[Domain],
               cfg: CFG).void }
  def initialize(direction, boundary, init, cfg)
    @direction = direction
    @boundary = boundary
    @init = init
    @cfg = cfg

    @out = T.let(Hash.new, T::Hash[BB, T::Set[Domain]])
    @in = T.let(Hash.new, T::Hash[BB, T::Set[Domain]])
    @gen = T.let(Hash.new, T::Hash[BB, T::Set[Domain]])
    @kill = T.let(Hash.new, T::Hash[BB, T::Set[Domain]])
  end

  sig { returns(CFGFacts[Domain]) }
  def run
    # run the dfa
    run_dfa

    # construct and return a CFGFacts object
    facts = CFGFacts.new(@cfg)
    facts.add_fact_hash(:out, @out)
    facts.add_fact_hash(:in, @in)
    facts.add_fact_hash(:gen, @gen)
    facts.add_fact_hash(:kill, @kill)

    return facts
  end

  protected

  sig { void }
  def run_dfa
    case @direction
    when Direction::Forwards then run_forwards
    when Direction::Backwards then run_backwards
    else T.absurd(@direction)
    end
  end

  sig { params(block: BB).returns(T::Set[Domain]) }
  def transfer(block)
    raise("Transfer function is unimplemented")
  end

  sig { params(block: BB).returns(T::Set[Domain]) }
  def meet(block)
    raise("Meet function is unimplemented")
  end

  sig { params(hash: T::Hash[BB, T::Set[Domain]], block: BB)
          .returns(T::Set[Domain]) }
  def get_set(hash, block)
    set_b = hash[block]

    if not set_b
      set_b = Set[]
    end

    return set_b
  end

  private

  sig { void }
  def run_forwards
    # initialize all nodes
    @out[@cfg.entry] = @boundary
    @cfg.each_block { |b|
      if b == @cfg.entry then next end

      @out[b] = @init
    }

    # iterate
    changed = T.let(true, T::Boolean)
    while changed
      changed = false
      @cfg.each_block { |b|
        if b == @cfg.entry then next end

        old_out = T.unsafe(@out[b])

        step_forwards(b)

        new_out = T.unsafe(@out[b])
        if old_out.length != new_out.length
          changed = true
        end
      }
    end
  end

  sig { params(block: BB).void }
  def step_forwards(block)
    @out[block] = transfer(block)
    @in[block] = meet(block)
  end

  sig { void }
  def run_backwards
    # initialize all nodes
    @in[@cfg.exit] = @boundary
    @cfg.each_block { |b|
      if b == @cfg.exit then next end

      @in[b] = @init
    }

    # iterate
    changed = T.let(true, T::Boolean)
    while changed
      changed = false
      @cfg.each_block { |b|
        if b == @cfg.exit then next end

        old_in = T.unsafe(@in[b])

        step_backwards(b)

        new_in = T.unsafe(@in[b])
        if old_in.length != new_in.length
          changed = true
        end
      }
    end
  end

  sig { params(block: BB).void }
  def step_backwards(block)
    @in[block] = transfer(block)
    @out[block] = meet(block)
  end
end

File: Lilac/src/Analysis/DFA.cs
namespace Lilac.Analysis;

abstract class DFA<T>(DFA<T>.Direction direction,
                      HashSet<T> boundary,
                      HashSet<T> @init,
                      CFG cfg) {
  public enum Direction {
    Forwards,
    Backwards,
  }

  private readonly Direction direction = direction;
  private readonly HashSet<T> boundary = boundary;
  private readonly HashSet<T> @init = @init;
  protected CFG CFG { get; } = cfg;

  protected Dictionary<BB, HashSet<T>> Out = [];
  protected Dictionary<BB, HashSet<T>> In = [];
  protected Dictionary<BB, HashSet<T>> Gen = [];
  protected Dictionary<BB, HashSet<T>> Kill = [];

  public CFGFacts<T> Run() {
    // run custom fact set initialization for each block
    foreach (BB n in CFG.GetNodes()) {
      InitSets(n);
    }

    // run the DFA
    switch (direction) {
      case Direction.Forwards:
        RunForwards();
        break;
      case Direction.Backwards:
        RunBackwards();
        break;
    }

    // construct and return the CFGFacts from the run
    CFGFacts<T> facts = new(CFG);
    facts.AddFactDict("out", Out);
    facts.AddFactDict("in", In);
    facts.AddFactDict("gen", Gen);
    facts.AddFactDict("kill", Kill);

    return facts;
  }

  protected abstract HashSet<T> Transfer(BB block);

  protected abstract HashSet<T> Meet(BB block);

  protected abstract void InitSets(BB block);

  protected static HashSet<T> GetSet(Dictionary<BB, HashSet<T>> dict, BB block) =>
    dict.TryGetValue(block, out HashSet<T>? blockSet)
      ? blockSet
      : [];

  private void RunForwards() {
    // initialize all nodes
    Out.Add(CFG.Entry, boundary);
    foreach (BB n in CFG.GetNodes().Where(n => n != CFG.Entry)) {
      Out.Add(n, @init);
    }

    // iterate
    bool changed = true;
    while (changed) {
      changed = false;

      foreach (BB n in CFG.GetNodes()) {
        if (n == CFG.Entry) {
          continue;
        }

        if (!Out.TryGetValue(n, out HashSet<T>? oldOut)) {
          throw new Exception(); // TODO: nicer exception type
        }

        StepForwards(n);

        if (!Out.TryGetValue(n, out HashSet<T>? newOut)) {
          throw new Exception(); // TODO: nicer exception type
        }

        if (newOut.Count != oldOut.Count) {
          changed = true;
        }
      }
    }
  }

  private void StepForwards(BB block) {
    In[block] = Meet(block);
    Out[block] = Transfer(block);
  }

  private void RunBackwards() {
    // initialize all nodes
    In.Add(CFG.Exit, boundary);
    foreach (BB n in CFG.GetNodes().Where(n => n != CFG.Exit)) {
      In.Add(n, @init);
    }

    // iterate
    bool changed = true;
    while (changed) {
      changed = false;

      foreach (BB n in CFG.GetNodes()) {
        if (n == CFG.Exit) {
          continue;
        }

        if (!In.TryGetValue(n, out HashSet<T>? oldIn)) {
          throw new Exception(); // TODO: nicer exception type
        }

        StepBackwards(n);

        if (!In.TryGetValue(n, out HashSet<T>? newIn)) {
          throw new Exception(); // TODO: nicer exception type
        }

        if (newIn.Count != oldIn.Count) {
          changed = true;
        }
      }
    }
  }

  private void StepBackwards(BB block) {
    Out[block] = Meet(block);
    In[block] = Transfer(block);
  }
}
